### ISSUE TITLE
Support for Synaptics 06cb:00b7 (Sensor Type 0xd51)

### ISSUE THREAD

**Issue by rsomonte** [uunicorn/python-validity#225]
Hello, I'm trying to get a fingerprint sensor with USB ID 06cb:00b7 working. After extensive troubleshooting, it appears the device is unsupported but very close to being compatible. The initialization fails even after applying several patches to the driver source code.

Here is a summary of the manual patches applied in order:

1. Udev Rule: Edited /usr/lib/udev/rules.d/60-python-validity.rules to add a line for the 00b7 product ID to ensure the python3-validity.service is triggered for this 

2. Device ID in Driver: Edited /usr/lib/python3.13/site-packages/validitysensor/usb.py to add DEV_b7 = (0x06cb, 0x00b7) to the SupportedDevices Enum, which fixed an initial "No matching devices found" error.

3. Firmware Blob Loading: Edited /usr/lib/python3.13/site-packages/validitysensor/blobs.py. The __load_blob function was failing with an UnboundLocalError because it did not have a case for the 00b7 product ID. I modified the line if usb.usb_dev().idProduct == 0x009a: to if usb.usb_dev().idProduct == 0x009a or usb.usb_dev().idProduct == 0x00b7: to make my device load the blobs_9a module.

4. Sensor Type Profile: Edited /usr/lib/python3.13/site-packages/validitysensor/sensor.py. In the open function, the sensor was correctly identified as sensor type 0xd51 but was then rejected. I modified the line if self.device_info.type == 0x199: to if self.device_info.type == 0x199 or self.device_info.type == 0xd51: to make the driver use the 0x199 sensor profile.

5. Capture Program: Still in sensor.py, the driver then failed with the error Can't find initial capture program. I modified the line self.hardcoded_prog = SensorCaptureProg.get(self.rom_info, self.device_info.type, ...) to self.hardcoded_prog = SensorCaptureProg.get(self.rom_info, 0x199, ...) to force it to load the capture program for the supported 0x199 sensor.

After this final change, the python3-validity.service still fails to start. The full journalctl log for this final failure is attached below.

lsusb output:

> Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
> Bus 001 Device 002: ID 04d9:fc30 Holtek Semiconductor, Inc. Gaming Mouse [Redragon M711]
> Bus 001 Device 003: ID 06cb:00b7 Synaptics, Inc. Fingerprint reader [HP G6]
> Bus 001 Device 004: ID 04f2:b669 Chicony Electronics Co., Ltd HP HD Camera
> Bus 001 Device 005: ID 8087:0029 Intel Corp. AX200 Bluetooth
> Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub

[journalctllog.txt](https://github.com/user-attachments/files/21445998/Documentsjournalctllog.txt)

**Comment by SimpleX-T** [comment]
Hey @rsomonte — installation is now packaged so you can drop the manual patches:

```bash
sudo add-apt-repository ppa:devtochukwu/fingerprint
sudo apt install python3-validity
```

Builds for Ubuntu 22.04, 24.04, 25.10, and 26.04. Let me know if you hit anything on your machine.

**Comment by SimpleX-T** [comment]
Progress update: `06cb:00b7` support has been iterated on in PR #256 through several rounds.

- Initial 0xd51 handling covers the HP G6 variants that report sensor type 0xd51 (@RamcharanRajpurohit confirmed working on Fedora 43).
- @Karloss1234's Point 1 review added 0x969 support (the other silicon behind this same USB ID, reporting sensor type 0x969 — HP ZBook 17 G6, ZBook Studio G5, ProBook G6 etc.). Merged in `0c00f0a`.
- His Point 4 fix for post-suspend/resume regression on 0x969 chips (the chip re-enumerates reporting 0x199 directly rather than 0x969, so `Sensor.open()` was skipping the alias + interrupt fix). Landed in `3b23194`.
- His Point 6 diagnosis of flash-management (`04b5` on enroll, `del_record` missing `db_write_enable`, no auto-compaction) — the `del_record` bug is fixed in `82584e1`; the "chip doesn't auto-compact after deletion" facet is documented with `erase_flash(4)` as recovery.

Karloss's review of PR #256 is the reference thread for the 0x969 details on this chip family. Full end-to-end on Kubuntu 26.04 + HP ZBook 17 G6: enroll, verify, sudo, KDE lock screen, resume from suspend.

If you're on 0x969 hardware and can test after the next PPA build (`0.16~hp8`, incoming), that would help close this out.

**Comment by rsomonte** [comment]
### Device Support Confirmation: HP ProBook 650 G7 (06cb:00b7 / Sensor Type 0xd51)                                                                                                                             
                                                                                                                                                                                                                 
  Hello, I can confirm that the  06cb:00b7  Synaptics fingerprint reader (Sensor Type  0xd51 ) is working perfectly on my HP ProBook 650 G7 using the changes in PR #256 (from branch  feat/sensor-type-0xd51  on
  SimpleX-T/python-validity ).                                                                                                                                                                                   
                                                                                                                                                                                                                 
  Here are the details of the testing:                                                                                                                                                                           
                                                                                                                                                                                                                 
  1. **Firmware Extraction**: Successful. Running  validity-sensors-firmware  downloaded and extracted  6_07f_hp_cmit_mis_qm.xpfwext  from the HP Softpaq package  sp135736.exe .                                    
  2. **Enrolling**: ` fprintd-enroll`  runs, communicates correctly with the device, and completes enrollment:                                                                                                         
```
    Using device /net/reactivated/Fprint/Device/0                                                                                                                                                                
    Enrolling right-index-finger finger.                                                                                                                                                                         
    Enroll result: enroll-stage-passed                                                                                                                                                                           
    Enroll result: enroll-retry-scan                                                                                                                                                                             
    Enroll result: enroll-stage-passed                                                                                                                                                                           
    Enroll result: enroll-stage-passed                                                                                                                                                                           
    Enroll result: enroll-stage-passed                                                                                                                                                                           
    Enroll result: enroll-stage-passed                                                                                                                                                                           
    Enroll result: enroll-completed      
```                                                                                                                                                                        
                                                                                                                                                                                                                 
  3. **Authentication**: Works flawlessly for system authentication (PAM / GDM /  sudo ).                                                                                                                            
                                                                                                                                                                                                                 
  #### System Information ( fastfetch )                                                                                                                                                                          
                                                                                                                                                                                                                 
    OS: Fedora Linux 44 (Workstation Edition) x86_64                                                                                                                                                             
    Host: HP ProBook 650 G7 Notebook PC (SBKPFV3)                                                                                                                                                                
    Kernel: Linux 7.0.14-201.fc44.x86_64                                                                                                                                                                         
    DE: GNOME 50.3                                                                                                                                                                                               
    WM: Mutter (Wayland)                                                                                                                                                                                         
    CPU: Intel(R) Core(TM) i5-10310U (8) @ 4.40 GHz                                                                                                                                                              
    Memory: 31.11 GiB                                                                                                                                                                                            
                                                                                                                                                                                                                 
  #### DMI System Hardware Info ( dmidecode )                                                                                                                                                                    
  
    System Information
        Manufacturer: HP
        Product Name: HP ProBook 650 G7 Notebook PC
        Version: SBKPFV3
        SKU Number: 166J4AV
        Family: 103C_5336AN HP ProBook
  
  #### USB Device Info ( lsusb -v )
  
    Bus 001 Device 002: ID 06cb:00b7 Synaptics, Inc. Fingerprint reader [HP G6]
    Device Descriptor:
      bLength                18
      bDescriptorType         1
      bcdUSB               2.00
      bDeviceClass          255 Vendor Specific Class
      bDeviceSubClass        16 [unknown]
      bDeviceProtocol       255 
      bMaxPacketSize0         8
      idVendor           0x06cb Synaptics, Inc.
      idProduct          0x00b7 Fingerprint reader [HP G6]
      bcdDevice            1.64
      iManufacturer           0 
      iProduct                0 
      iSerial                 1 55779486f5d0
      bNumConfigurations      1
      Configuration Descriptor:
        bLength                 9
        bDescriptorType         2
        wTotalLength       0x0035
        bNumInterfaces          1
        bConfigurationValue     1
        iConfiguration          0 
        bmAttributes         0xa0
          (Bus Powered)
          Remote Wakeup
        MaxPower              100mA

**Comment by SimpleX-T** [comment]
Thanks @rsomonte — this is exactly the kind of validation we needed for this issue's specific `06cb:00b7 + 0xd51` combo, on a model (HP ProBook 650 G7) that hadn't been reported before. Fedora 44 + kernel 7.0.14 + GNOME 50.3 Wayland is also new ground.

For anyone else reading this on the same USB ID: `06cb:00b7` has been observed as **two different silicon variants** in the wild:
- `0xd51` — HP G6 series (this reporter's ProBook 650 G7, plus RamcharanRajpurohit's earlier Fedora 43 report, plus @bcoutts's ProBook 445R G6 — although his is stuck on a separate factory-init issue tracked on PR #256).
- `0x969` — HP ZBook 17 G6 (Karloss1234's report). Same PID, different chip, added to the PR in `0c00f0a` from @ggiesen's [SimpleX-T/python-validity#2](https://github.com/SimpleX-T/python-validity/pull/2).

Both variants are handled on the current PR #256 branch. `0.16~hp8` was uploaded to the PPA last night rolling up all the recent fixes — Launchpad's build queue should have the Fedora-equivalents (well, the Ubuntu builds — Fedora users would still install from source via COPR / setup.py, since we don't ship a `.spec` here).

If you have the time, a `bench_capture.py` run on your machine would give us a second empirical data point on capture quality for this chip family (my baseline is ~0.9 retries/verify on the 840 G5, 0xd51). The harness is in the source tree at `bench_capture.py`; runs 10 back-to-back `fprintd-verify` invocations and reports mean/median/p95 chip-side retries. Purely optional though — the enroll + PAM confirmation you posted is already the important bit.

**Comment by rsomonte** [comment]
### Capture Quality Benchmark & Feedback (HP ProBook 650 G7 / 06cb:00b7 / 0xd51)
 
Here are the capture quality metrics from my machine.
 
Note that I couldn't find the `bench_capture.py` script in the source tree of the repository, so I wrote a custom harness script that runs `fprintd-verify` 10 times and parses the `journalctl -u python3-validity.service` log output to calculate the stats based on the `Chip capture retry-scan #X` entries.
 
Here are the results:

| Run | Retries | Match Outcome |
| :--- | :--- | :--- |
| Run 1 | 0 | `verify-no-match` |
| Run 2 | 3 | `verify-no-match` |
| Run 3 | 0 | `verify-no-match` |
| Run 4 | 0 | `verify-no-match` |
| Run 5 | 0 | `verify-match` |
| Run 6 | 0 | `verify-no-match` |
| Run 7 | 0 | `verify-no-match` |
| Run 8 | 1 | `verify-match` |
| Run 9 | 0 | `verify-no-match` |
| Run 10 | 1 | `verify-no-match` |
| **Mean** | **0.50** | |
| **Median** | **0.0** | |
| **p95** | **3.0** | |
 
#### Summary & Feedback
 
* **Mean Retries:** 0.50 (very fast and clean compared to the G5 baseline of ~0.9).
* **Median:** 0.0
* **p95:** 3.0
* **Matching Observation:** During the test runs, even when using the correct enrolled finger, the verification returned `verify-no-match` on 8 out of 10 attempts (only matching successfully on Run 5 and Run 8). It seems that while the capture itself is extremely fast with low retry counts, the matching threshold is quite sensitive, and slightly off-angle scans are registered as incorrect fingerprints rather than asking for a retry. Re-enrolling or scanning at the exact angle helps, but there is a noticeable rate of false negatives.

**Comment by Karloss1234** [comment]
Quote:
``
     * Matching Observation:** During the test runs, even when using the correct enrolled finger, the verification returned `verify-no-match` on 8 out of 10 attempts (only matching successfully on Run 5 and Run 8). It seems that while the capture itself is extremely fast with low retry counts, the matching threshold is quite sensitive, and slightly off-angle scans are registered as incorrect fingerprints rather than asking for a retry. Re-enrolling or scanning at the exact angle helps, but there is a noticeable rate of false negatives.
 *
``

Does your system have a separate boot loaded Windows OS, with a finger enrolled in Windows Hello?

I had similar results to you when I had my right index finger enrolled in both Windows Hello and Linux fprintd-enrol.
In Linux fprintd-verify, all stored templates compete for verification and Windows Hello enroll template algorithm seems to be better at verifying, therefore the Windows Hello template wins fprintd-verify 80-90% of attempts, and therefore you get a 'verify-no-match'.
Delete/Switch your Windows Hello enrolled finger to your middle finger and see if that helps.  It resulted in me getting a 95% Linux verify-match rate.

**Comment by rsomonte** [comment]
Your diagnosis was 100% correct!
  
  After switching the Windows Hello enrolled finger to a different finger and re-enrolling my right index finger under Linux `fprintd-enroll`, the template competition issue completely disappeared.            
  
  I re-ran the 10-invocation test harness, and all 10 out of 10 runs matched successfully on the first attempt (100% match rate).
  
  #### Benchmark Results
  
| Run    | Retries | Outcome                     |
|---------|---------:|-----------------------------|
| Run 1   | 6        | verify-match (exit 0)       |
| Run 2   | 0        | verify-match (exit 0)       |
| Run 3   | 1        | verify-match (exit 0)       |
| Run 4   | 0        | verify-match (exit 0)       |
| Run 5   | 1        | verify-match (exit 0)       |
| Run 6   | 0        | verify-match (exit 0)       |
| Run 7   | 0        | verify-match (exit 0)       |
| Run 8   | 0        | verify-match (exit 0)       |
| Run 9   | 0        | verify-match (exit 0)       |
| Run 10  | 0        | verify-match (exit 0)       |
| **Mean**   | **0.80** | **100% Match Success** |
| **Median** | **0.0**  |                         |
| **p95**    | **6.0**  |                         |
  
  #### Summary & Takeaways
  
  • **Match Success Rate:** 100% (10/10 successful `verify-match`).
  • **Mean Retries**: 0.80 (7 out of 10 scans matched on 0 retries).
  • **Root Cause Confirmation:** Having the exact same finger enrolled in both Windows Hello and Linux `fprintd` causes the on-chip Windows Hello template to win the match search ~80% of the time, causing Linux   
 `fprintd-verify` to see a SID mismatch and return `verify-no-match`.
  • **Recommendation:** Dual-boot users on 0xd51 / 0x969 silicon should ensure they do not enroll the exact same finger in both Windows Hello and Linux  fprintd , or assign separate fingers for each OS.
